src/mlpack/methods/reinforcement_learning/eql.hpp

```diff
@@ -0,0 +1,194 @@
+/**
+ * @file methods/reinforcement_learning/eql.hpp
+ * @author Nanubala Gnana Sai
+ *
+ * This file is implements a Multi-objective Q learning algorithm,
+ * which is called Envelope Q-learning.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_EQL_HPP
+#define MLPACK_METHODS_RL_EQL_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <ensmallen.hpp>
+
+#include "replay/random_replay.hpp"
+#include "replay/prioritized_replay.hpp"
+#include "training_config.hpp"
+
+namespace mlpack {
+namespace rl {
+
+/**
+ * Implementation of EQL.
+ *
+ * For more details, see the following:
+ * @code
+ * @article{yang2019generalized,
+ *  author    = {Yang and
+ *               Runzhe and
+ *               Sun and
+ *               Xingyuan and
+ *               Narasimhan and
+ *               Karthik},
+ *  title     = {A generalized algorithm for multi-objective reinforcement learning and policy adaptation},
+ *  journal   = {arXiv preprint arXiv:1908.08342},
+ *  year      = {2019},
+ *  url       = {https://arxiv.org/abs/1908.08342}
+ * }
+ * @endcode
+ *
+ * @tparam EnvironmentType The environment of the reinforcement learning task.
+ * @tparam NetworkType The network to compute utility value.
+ * @tparam UpdaterType How to apply gradients when training.
+ * @tparam PolicyType Behavior policy of the agent.
+ * @tparam ReplayType Experience replay method.
+ */
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType = PrioritizedReplay<EnvironmentType>
+>
+class EQL
+{
```

Comment:
I want to set the distribution of the preference vector, how do I do it? For ex: whether preference weights are sampled from gaussian or gamma distribution during training.  The D_w part in the paper

![image](https://user-images.githubusercontent.com/45007169/128625263-6a6cc6f3-93e3-4290-9781-911a70b8a298.png)

Reply:
In the *policy adaptation* part just after the algorithm definition, they mention using a multivariate gaussian with m means and a single standard deviation to sample the preference vector from. However, I am not quite sure if they use the same distribution for learning too.

![IMG_7F7ED162ED5D-1](https://user-images.githubusercontent.com/8839504/128625849-b1a99af6-3ef8-4b7d-88e0-08f2a63cf110.jpeg)

Reply:
Yes true! The code agrees with you as well. Here's what they've done in code:

```python
        if preference is None:
            if self.w_kept is None:
                self.w_kept = torch.randn(self.model_.reward_size)
                self.w_kept = (torch.abs(self.w_kept) / \
                               torch.norm(self.w_kept, p=1)).type(FloatTensor)
            preference = self.w_kept
```

I guess this is the "assumption part", because preference distribution was ```None```.  We should equip users with the ability to set preference distribution.

Reply:
We could keep it as an optional argument and default to the multi-variate gaussian too; as the authors did. :)